File: suc/swift/024.swift

import SwiftUI

// MARK: - Color Extension for Hex Initialization
// 允许使用十六进制值初始化颜色，与 Kotlin 的 Color(0xFF...) 对应
extension Color {
    init(hex: UInt, alpha: Double = 1.0) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: alpha
        )
    }
}

// MARK: - AppTokens
// 对应 Kotlin 中的 AppTokens 对象，定义了颜色、字体、间距、形状和阴影
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFFEF476F)
        static let secondary = Color(hex: 0xFFFFD166)
        static let tertiary = Color(hex: 0xFF06D6A0)
        static let background = Color(hex: 0xFFFFFCF2)
        static let surface = Color(hex: 0xFFFFFFFF)
        static let surfaceVariant = Color(hex: 0xFFF7EDE2)
        static let outline = Color(hex: 0xFFE5D4B1)
        static let success = Color(hex: 0xFF06D6A0)
        static let warning = Color(hex: 0xFFFFA600)
        static let error = Color(hex: 0xFFD62828)
        static let onPrimary = Color(hex: 0xFFFFFFFF)
        static let onSecondary = Color(hex: 0xFF1E1E1E)
        static let onTertiary = Color(hex: 0xFF1E1E1E)
        static let onBackground = Color(hex: 0xFF1E1E1E)
        static let onSurface = Color(hex: 0xFF1E1E1E)
    }

    struct TypographyTokens {
        // 使用系统字体，直接映射 Compose 的 TextStyle
        static let display = Font.system(size: 26, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular)
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let smallCornerRadius: CGFloat = 6.0
        static let mediumCornerRadius: CGFloat = 10.0
        static let largeCornerRadius: CGFloat = 16.0
    }

    struct Spacing {
        static let xs: CGFloat = 4.0
        static let sm: CGFloat = 8.0
        static let md: CGFloat = 12.0
        static let lg: CGFloat = 16.0
        static let xl: CGFloat = 24.0
    }

    struct ShadowSpec {
        let elevation: CGFloat // 对应 SwiftUI shadow 的 blur radius
        let radius: CGFloat // 在 SwiftUI shadow 中不直接使用，elevation 承担主要作用
        let dy: CGFloat // 对应 SwiftUI shadow 的 y 偏移
        let opacity: Double
    }

    struct ElevationMapping {
        // Compose 的 elevation 映射到 SwiftUI shadow 的 radius (模糊半径) 和 y 偏移
        static let level1 = ShadowSpec(elevation: 2.0, radius: 4.0, dy: 2.0, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 4.0, radius: 8.0, dy: 4.0, opacity: 0.16)
    }
}

// MARK: - RootScreen 对应的 SwiftUI 视图
struct RootScreen: View {
    @State private var points: [CGPoint] = [] // 存储绘制点的数组
    @State private var currentColor: Color = AppTokens.Colors.primary // 当前绘制颜色
    
    // 定义控件尺寸，以实现原子级对应
    let buttonHeight: CGFloat = 44.0 // 底部颜色按钮的高度，根据 Android Material3 默认按钮文本大小和垂直内边距估算
    let fabSize: CGFloat = 56.0 // 标准 Material Design Floating Action Button 尺寸
    let circleRadius: CGFloat = 10.0 // 绘制圆点的半径

    // 计算顶部栏的实际高度，用于内容区域的内边距
    var topBarActualHeight: CGFloat {
        // 字体大小 26pt，近似行高 30pt。加上 2 倍垂直内边距 (AppTokens.Spacing.md = 12pt)
        return 30.0 + (2 * AppTokens.Spacing.md) // 30 + 2*12 = 54
    }
    
    // 计算底部栏的实际高度，用于内容区域的内边距
    var bottomBarActualHeight: CGFloat {
        // 按钮高度 (44pt) 加上 2 倍垂直内边距 (AppTokens.Spacing.md = 12pt)
        return buttonHeight + (2 * AppTokens.Spacing.md) // 44 + 2*12 = 68
    }
    
    // FAB 距离屏幕底部的内边距
    var fabBottomMargin: CGFloat {
        return AppTokens.Spacing.xl // 24pt
    }
    
    // 绘制画布的底部内边距，需考虑底部栏和 FAB 中占据空间较大的一个
    var canvasBottomPadding: CGFloat {
        // Compose Scaffold 的内容区域内边距会取底部栏高度和 FAB 底部边缘位置的最大值
        // FAB 底部边缘位置 = FAB 尺寸 + FAB 距离屏幕底部边距 = 56 + 24 = 80
        // 所以取 max(底部栏实际高度, FAB 底部边缘位置) = max(68, 80) = 80
        return max(bottomBarActualHeight, fabSize + fabBottomMargin)
    }
    
    var body: some View {
        ZStack {
            // 1. 背景色：填充整个屏幕，包括状态栏区域
            AppTokens.Colors.background
                .ignoresSafeArea()

            // 2. 绘制画布区域：用户在此区域进行绘制
            Canvas { context, size in
                for point in points {
                    let rect = CGRect(x: point.x - circleRadius, y: point.y - circleRadius, width: circleRadius * 2, height: circleRadius * 2)
                    context.fill(Path(ellipseIn: rect), with: .color(currentColor))
                }
            }
            // 拖动手势识别，用于在画布上添加点
            .gesture(
                DragGesture(minimumDistance: 0) // minimumDistance: 0 意味着即使是点击也会触发 onChanged
                    .onChanged { value in
                        // `value.location` 是相对于 Canvas 视图左上角的坐标。
                        // 由于 Canvas 视图本身被内边距推移，其坐标系已与 Compose 的 `padding(pad)` 后的 `pointerInput` 行为一致。
                        points.append(value.location)
                    }
            )
            // 为 Canvas 视图应用内边距，以避开顶部栏、底部栏和 FAB 占据的区域
            .padding(.top, topBarActualHeight)
            .padding(.bottom, canvasBottomPadding)

            // 3. 顶部栏
            VStack {
                Text("Retro Whiteboard")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.onSurface)
                    .padding(.vertical, AppTokens.Spacing.md) // 标题的垂直内边距
                    .frame(maxWidth: .infinity) // 使文本水平居中
                    .background(AppTokens.Colors.background) // 顶部栏背景色
            }
            .frame(maxHeight: .infinity, alignment: .top) // 将 VStack 对齐到 ZStack 的顶部

            // 4. 底部颜色选择栏
            VStack {
                Spacer() // 将 HStack 推到 ZStack 的底部
                HStack(spacing: AppTokens.Spacing.md) { // 按钮之间的间距
                    Spacer() // 用于实现 SpaceEvenly 效果
                    Button(action: { currentColor = AppTokens.Colors.primary }) {
                        Text("Red")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onPrimary)
                            .frame(minWidth: 0, maxWidth: .infinity) // 允许按钮水平填充可用空间
                            .frame(height: buttonHeight) // 固定按钮高度
                            .background(AppTokens.Colors.primary)
                            .cornerRadius(AppTokens.Shapes.smallCornerRadius)
                    }
                    Spacer()
                    Button(action: { currentColor = AppTokens.Colors.secondary }) {
                        Text("Yellow")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(minWidth: 0, maxWidth: .infinity)
                            .frame(height: buttonHeight)
                            .background(AppTokens.Colors.secondary)
                            .cornerRadius(AppTokens.Shapes.smallCornerRadius)
                    }
                    Spacer()
                    Button(action: { currentColor = AppTokens.Colors.tertiary }) {
                        Text("Green")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onTertiary)
                            .frame(minWidth: 0, maxWidth: .infinity)
                            .frame(height: buttonHeight)
                            .background(AppTokens.Colors.tertiary)
                            .cornerRadius(AppTokens.Shapes.smallCornerRadius)
                    }
                    Spacer()
                }
                .padding(AppTokens.Spacing.md) // 底部栏内部，按钮 HStack 的内边距
                .background(AppTokens.Colors.surface) // 底部栏背景色
                .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                        radius: AppTokens.ElevationMapping.level1.elevation, // 模糊半径
                        x: 0, y: AppTokens.ElevationMapping.level1.dy) // Y 轴偏移
            }
            .frame(maxHeight: .infinity, alignment: .bottom) // 将 VStack 对齐到 ZStack 的底部

            // 5. 浮动操作按钮 (Floating Action Button) - 清除画布
            VStack {
                Spacer() // 将 HStack 推到 ZStack 的底部
                HStack {
                    Spacer() // 将 Button 推到 HStack 的右侧 (trailing)
                    Button(action: { points.removeAll() }) { // 清除所有绘制点
                        Text("Clear")
                            .font(AppTokens.TypographyTokens.label)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(width: fabSize, height: fabSize) // 固定 FAB 尺寸
                            .background(AppTokens.Colors.secondary)
                            .clipShape(Circle()) // 圆形剪裁
                            .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                                    radius: AppTokens.ElevationMapping.level1.elevation,
                                    x: 0, y: AppTokens.ElevationMapping.level1.dy)
                    }
                    .padding(.trailing, AppTokens.Spacing.xl) // FAB 右侧内边距
                    .padding(.bottom, fabBottomMargin) // FAB 底部内边距
                }
            }
            .frame(maxHeight: .infinity, alignment: .bottom) // 将 VStack 对齐到 ZStack 的底部
        }
        .statusBarHidden(true) // 隐藏整个视图的状态栏 (iOS 16.0 兼容)
    }
}

// MARK: - 主应用结构
// @main 入口点，使应用可直接运行
@main
struct RetroWhiteboardApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}

File: suc/kotlin/024.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.compose.foundation.gestures.detectDragGestures

private const val NAME = "025Whiteboarden"
private const val UI_TYPE = "Productivity"
private const val STYLE_THEME = "Retro Flat"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFEF476F)
        val secondary = Color(0xFFFFD166)
        val tertiary = Color(0xFF06D6A0)
        val background = Color(0xFFFFFCF2)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFF7EDE2)
        val outline = Color(0xFFE5D4B1)
        val success = Color(0xFF06D6A0)
        val warning = Color(0xFFFFA600)
        val error = Color(0xFFD62828)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1E1E1E)
        val onTertiary = Color(0xFF1E1E1E)
        val onBackground = Color(0xFF1E1E1E)
        val onSurface = Color(0xFF1E1E1E)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(4.dp, 8.dp, 4.dp, 0.16f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    val points = remember { mutableStateListOf<Offset>() }
    var color by remember { mutableStateOf(AppTokens.Colors.primary) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        floatingActionButton = {
            FloatingActionButton(
                onClick = { points.clear() },
                containerColor = AppTokens.Colors.secondary,
                contentColor = AppTokens.Colors.onSecondary,
                shape = CircleShape
            ) { Text("Clear", style = MaterialTheme.typography.labelMedium) }
        },
        bottomBar = {
            Surface(
                tonalElevation = AppTokens.ElevationMapping.level1.elevation,
                shadowElevation = AppTokens.ElevationMapping.level1.elevation,
                color = AppTokens.Colors.surface
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(AppTokens.Spacing.md),
                    horizontalArrangement = Arrangement.SpaceEvenly,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Button(
                        onClick = { color = AppTokens.Colors.primary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                        shape = AppTokens.Shapes.small
                    ) { Text("Red") }
                    Button(
                        onClick = { color = AppTokens.Colors.secondary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary, contentColor = AppTokens.Colors.onSecondary),
                        shape = AppTokens.Shapes.small
                    ) { Text("Yellow") }
                    Button(
                        onClick = { color = AppTokens.Colors.tertiary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.tertiary, contentColor = AppTokens.Colors.onTertiary),
                        shape = AppTokens.Shapes.small
                    ) { Text("Green") }
                }
            }
        },
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Retro Whiteboard", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.onSurface) },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(AppTokens.Colors.background)
                .pointerInput(Unit) {
                    detectDragGestures { change, _ ->
                        points.add(change.position)
                    }
                }
        ) {
            Canvas(modifier = Modifier.fillMaxSize()) {
                points.forEach {
                    drawCircle(color = color, radius = 10f, center = it)
                }
            }
        }
    }


}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFFCF2)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}